# Task 1.2 - Scaling ClimateWins Temperature Dataset

## Table of Contents

#### 1. Importing Libraries and Data
- Original dataset from ClimateWins

#### 2. Checking for Missing Values

#### 3. Scaling Data

#### 4. Exporting Scaled Dataframe

## 1. Importing Libraries and Data

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib
import os
import sklearn
from sklearn.preprocessing import StandardScaler

In [2]:
path = r'/Users/giadairene/Documents/CareerFoundry Data Analytics/Machine Learning with Python/ClimateWins/02 Data'

In [3]:
# Import data

df = pd.read_csv(os.path.join(path, 'Original/Dataset-weather-prediction-dataset-processed.csv'))

In [4]:
df.shape

(22950, 170)

In [5]:
# Delimit columns displayed

pd.options.display.max_columns = None

In [6]:
df.head(10)

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0  19600101      1                  7               2.1            0.85   
1  19600102      1                  6               2.1            0.84   
2  19600103      1                  8               2.1            0.90   
3  19600104      1                  3               2.1            0.92   
4  19600105      1                  6               2.1            0.95   
5  19600106      1                  1               2.1            0.78   
6  19600107      1                  3               2.1            0.87   
7  19600108      1                  8               2.1            0.95   
8  19600109      1                  8               2.1            0.98   
9  19600110      1                  3               2.1            0.74   

   BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0           1.018                    0.32                 0.09   
1           1.018                    0.36                 1.05   
2           1.018                    0.18                 0.30   
3           1.018                    0.58                 0.00   
4           1.018                    0.65                 0.14   
5           1.018                    0.76                 0.00   
6           1.018                    0.66                 0.00   
7           1.018                    0.19                 0.09   
8           1.018                    0.19                 0.56   
9           1.018                    0.38                 0.01   

   BASEL_snow_depth  BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  \
0                 0             0.7              6.5             0.8   
1                 0             1.1              6.1             3.3   
2                 0             0.0              8.5             5.1   
3                 0             4.1              6.3             3.8   
4                 0             5.4              3.0            -0.7   
5                 0             7.7              1.5            -1.8   
6                 0             5.4             -0.4            -4.5   
7                 0             0.0              0.9            -2.2   
8                 0             0.0             -2.4            -4.9   
9                 5             1.1             -6.9           -10.6   

   BASEL_temp_max  BELGRADE_cloud_cover  BELGRADE_humidity  BELGRADE_pressure  \
0            10.9                     1               0.81             1.0195   
1            10.1                     6               0.84             1.0172   
2             9.9                     6               0.77             1.0179   
3            10.6                     8               0.93             1.0268   
4             6.0                     8               0.99             1.0286   
5             6.4                     8               0.84             1.0179   
6             6.1                     8               0.92             1.0141   
7             2.6                     6               0.93             1.0149   
8             1.5                     8               0.71             1.0082   
9            -3.4                     8               0.92             1.0136   

   BELGRADE_global_radiation  BELGRADE_precipitation  BELGRADE_sunshine  \
0                       0.88                    0.00                7.0   
1                       0.25                    0.00                0.0   
2                       0.67                    0.00                3.5   
3                       0.25                    0.00                0.0   
4                       0.25                    0.06                0.0   
5                       0.61                    0.64                2.5   
6                       0.25                    0.35                0.0   
7                       0.26                    0.14                0.0   
8                       0.49                    0.46                1.2   
9                  

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22950 entries, 0 to 22949
Columns: 170 entries, DATE to VALENTIA_temp_max
dtypes: float64(145), int64(25)
memory usage: 29.8 MB


In [21]:
# Set pandas options to display all rows

pd.options.display.max_rows = None

In [22]:
df['DATE'].value_counts()

DATE
19600101    1
20011106    1
20011126    1
20011125    1
20011124    1
20011123    1
20011122    1
20011121    1
20011120    1
20011119    1
20011118    1
20011117    1
20011116    1
20011115    1
20011114    1
20011113    1
20011112    1
20011111    1
20011110    1
20011109    1
20011108    1
20011127    1
20011128    1
20011129    1
20011210    1
20011218    1
20011217    1
20011216    1
20011215    1
20011214    1
20011213    1
20011212    1
20011211    1
20011209    1
20011130    1
20011208    1
20011207    1
20011206    1
20011205    1
20011204    1
20011203    1
20011202    1
20011201    1
20011107    1
20011105    1
20011220    1
20011104    1
20011012    1
20011011    1
20011010    1
20011009    1
20011008    1
20011007    1
20011006    1
20011005    1
20011004    1
20011003    1
20011002    1
20011001    1
20010930    1
20010929    1
20010928    1
20010927    1
20010926    1
20010925    1
20010924    1
20011013    1
20011014    1
20011015    1
20011026    1
20011103    1
2

In [24]:
# Filter columns that contain '_temp_mean' in their names

temp_mean_columns = df.filter(like='_temp_mean')

In [26]:
temp_mean_columns.max().sort_values(ascending=False)

BELGRADE_temp_mean      34.6
MADRID_temp_mean        33.4
BUDAPEST_temp_mean      33.1
ROMA_temp_mean          31.5
DUSSELDORF_temp_mean    31.2
TOURS_temp_mean         31.2
MAASTRICHT_temp_mean    30.9
HEATHROW_temp_mean      30.9
LJUBLJANA_temp_mean     30.4
DEBILT_temp_mean        29.7
MUNCHENB_temp_mean      29.5
BASEL_temp_mean         29.2
KASSEL_temp_mean        28.4
STOCKHOLM_temp_mean     28.3
GDANSK_temp_mean        27.6
OSLO_temp_mean          26.4
VALENTIA_temp_mean      23.6
SONNBLICK_temp_mean     12.7
dtype: float64

#### The highest average temperature ever registered in a day was in Belgrade, with 34.6°C.

In [28]:
temp_mean_columns.min().sort_values(ascending=True)

SONNBLICK_temp_mean    -31.8
STOCKHOLM_temp_mean    -23.9
GDANSK_temp_mean       -23.8
MUNCHENB_temp_mean     -22.5
OSLO_temp_mean         -22.0
BASEL_temp_mean        -18.5
BELGRADE_temp_mean     -17.3
BUDAPEST_temp_mean     -16.5
KASSEL_temp_mean       -15.8
LJUBLJANA_temp_mean    -15.5
MAASTRICHT_temp_mean   -14.9
DUSSELDORF_temp_mean   -14.6
DEBILT_temp_mean       -13.2
TOURS_temp_mean        -12.6
HEATHROW_temp_mean      -8.9
ROMA_temp_mean          -5.6
MADRID_temp_mean        -4.5
VALENTIA_temp_mean      -3.5
dtype: float64

#### The lowest average temperature ever registered in a day was in Sonnblick, with -31.8°C.

## 2. Checking for Missing Values

In [9]:
df.isnull().sum()

DATE                   0
MONTH                  0
BASEL_cloud_cover      0
BASEL_wind_speed       0
BASEL_humidity         0
                      ..
VALENTIA_snow_depth    0
VALENTIA_sunshine      0
VALENTIA_temp_mean     0
VALENTIA_temp_min      0
VALENTIA_temp_max      0
Length: 170, dtype: int64

In [10]:
# Delete records with NaN values

df.dropna(inplace=True)

In [11]:
df.shape

(22950, 170)

#### No missing values were found.

## 3. Scaling Data

In [12]:
# Create scaler object
# StandardScaler assumes data is normally distributed and scales with a distribution around 0
# and standard deviation of 1. Scaling happens independently with each variable.

scaler = StandardScaler()

In [13]:
# Create a new df with scaled data

df_scaled = pd.DataFrame(scaler.fit_transform(df), columns = list(df))

df_scaled.head()

DATE     MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0 -1.707663 -1.599964           0.660514          -0.02793        0.826097   
1 -1.707657 -1.599964           0.244897          -0.02793        0.735760   
2 -1.707652 -1.599964           1.076130          -0.02793        1.277781   
3 -1.707646 -1.599964          -1.001953          -0.02793        1.458455   
4 -1.707641 -1.599964           0.244897          -0.02793        1.729466   

   BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0       -0.001949               -1.101066            -0.265148   
1       -0.001949               -1.058108             1.658760   
2       -0.001949               -1.251420             0.155707   
3       -0.001949               -0.821838            -0.445514   
4       -0.001949               -0.746661            -0.164944   

   BASEL_snow_depth  BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  \
0         -0.179228       -0.902918        -0.528623       -0.845652   
1         -0.179228       -0.810126        -0.582946       -0.462450   
2         -0.179228       -1.065304        -0.257010       -0.186545   
3         -0.179228       -0.114186        -0.555784       -0.385810   
4         -0.179228        0.187388        -1.003946       -1.075573   

   BASEL_temp_max  BELGRADE_cloud_cover  BELGRADE_humidity  BELGRADE_pressure  \
0       -0.478356             -1.206433           0.905270           0.321771   
1       -0.569988              0.652846           1.119118           0.016838   
2       -0.592896              0.652846           0.620141           0.109644   
3       -0.512718              1.396557           1.760659           1.289601   
4       -1.039603              1.396557           2.188353           1.528244   

   BELGRADE_global_radiation  BELGRADE_precipitation  BELGRADE_sunshine  \
0                  -0.743149               -0.371315           0.283983   
1                  -1.418387               -0.371315          -1.333204   
2                  -0.968228               -0.371315          -0.524611   
3                  -1.418387               -0.371315          -1.333204   
4                  -1.418387               -0.253655          -1.333204   

   BELGRADE_temp_mean  BELGRADE_temp_min  BELGRADE_temp_max  \
0           -1.016876          -1.220210          -0.949203   
1           -1.107669          -0.818702          -1.298119   
2           -1.084971          -1.168403          -1.098738   
3           -1.209812          -1.362681          -1.437685   
4           -1.209812          -1.012980          -1.457623   

   BUDAPEST_cloud_cover  BUDAPEST_humidity  BUDAPEST_pressure  \
0              -0.25524          -0.010098          -0.005876   
1              -0.25524          -0.010098          -0.005876   
2              -0.25524          -0.010098          -0.005876   
3              -0.25524          -0.010098          -0.005876   
4              -0.25524          -0.010098          -0.005876   

   BUDAPEST_global_radiation  BUDAPEST_precipitation  BUDAPEST_sunshine  \
0                  -1.118537               -0.321868          -0.773378   
1                  -1.415938                0.371622          -1.344065   
2                  -1.278676               -0.344984          -1.195190   
3                  -1.404499               -0.344984          -1.344065   
4                  -1.404499               -0.344984          -1.344065   

   BUDAPEST_temp_mean  BUDAPEST_temp_min  BUDAPEST_temp_max  \
0           -1.099163          -1.119431          -1.136839   
1           -1.110927          -0.878727          -1.341204   
2           -1.063873          -0.838610          -1.116403   
3           -1.146217          -1.012452          -1.208367   
4           -1.087400          -0.918844          -1.116403   

   DEBILT_cloud_cover  DEBILT_wind_speed  DEBILT_humidity  DEBILT_pressure  \
0            0.713782           2.781340         0.357839        -1.232471   
1            1

In [14]:
# Compare the original vs. scaled data

df.head()

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0  19600101      1                  7               2.1            0.85   
1  19600102      1                  6               2.1            0.84   
2  19600103      1                  8               2.1            0.90   
3  19600104      1                  3               2.1            0.92   
4  19600105      1                  6               2.1            0.95   

   BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0           1.018                    0.32                 0.09   
1           1.018                    0.36                 1.05   
2           1.018                    0.18                 0.30   
3           1.018                    0.58                 0.00   
4           1.018                    0.65                 0.14   

   BASEL_snow_depth  BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  \
0                 0             0.7              6.5             0.8   
1                 0             1.1              6.1             3.3   
2                 0             0.0              8.5             5.1   
3                 0             4.1              6.3             3.8   
4                 0             5.4              3.0            -0.7   

   BASEL_temp_max  BELGRADE_cloud_cover  BELGRADE_humidity  BELGRADE_pressure  \
0            10.9                     1               0.81             1.0195   
1            10.1                     6               0.84             1.0172   
2             9.9                     6               0.77             1.0179   
3            10.6                     8               0.93             1.0268   
4             6.0                     8               0.99             1.0286   

   BELGRADE_global_radiation  BELGRADE_precipitation  BELGRADE_sunshine  \
0                       0.88                    0.00                7.0   
1                       0.25                    0.00                0.0   
2                       0.67                    0.00                3.5   
3                       0.25                    0.00                0.0   
4                       0.25                    0.06                0.0   

   BELGRADE_temp_mean  BELGRADE_temp_min  BELGRADE_temp_max  \
0                 3.7               -0.9                7.9   
1                 2.9                2.2                4.4   
2                 3.1               -0.5                6.4   
3                 2.0               -2.0                3.0   
4                 2.0                0.7                2.8   

   BUDAPEST_cloud_cover  BUDAPEST_humidity  BUDAPEST_pressure  \
0                     4               0.67              1.017   
1                     4               0.67              1.017   
2                     4               0.67              1.017   
3                     4               0.67              1.017   
4                     4               0.67              1.017   

   BUDAPEST_global_radiation  BUDAPEST_precipitation  BUDAPEST_sunshine  \
0                       0.44                    0.01                2.3   
1                       0.18                    0.31                0.0   
2                       0.30                    0.00                0.6   
3                       0.19                    0.00                0.0   
4                       0.19                    0.00                0.0   

   BUDAPEST_temp_mean  BUDAPEST_temp_min  BUDAPEST_temp_max  \
0                 2.4               -0.4                5.1   
1                 2.3                1.4                3.1   
2                 2.7                1.7                5.3   
3                 2.0                0.4                4.4   
4                 2.5                1.1                5.3   

   DEBILT_cloud_cover  DEBILT_wind_speed  DEBILT_humidity  DEBILT_pressure  \
0                   7                7.7             0.85           1.0032   
1                   8           

#### We should remove scaled Date and Month columns from the scaled dataset (as they represent dates, not numbers), and then attach the original Date and Month columns to prevent loss of information.

In [15]:
# Create a copy of the DATE and MONTH columns

date_month_columns = df[['DATE', 'MONTH']]

In [16]:
# Drop unnecessary scaled columns

df_scaled.drop(['DATE', 'MONTH'], axis=1, inplace=True)

In [17]:
df_scaled.head()

BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  BASEL_pressure  \
0           0.660514          -0.02793        0.826097       -0.001949   
1           0.244897          -0.02793        0.735760       -0.001949   
2           1.076130          -0.02793        1.277781       -0.001949   
3          -1.001953          -0.02793        1.458455       -0.001949   
4           0.244897          -0.02793        1.729466       -0.001949   

   BASEL_global_radiation  BASEL_precipitation  BASEL_snow_depth  \
0               -1.101066            -0.265148         -0.179228   
1               -1.058108             1.658760         -0.179228   
2               -1.251420             0.155707         -0.179228   
3               -0.821838            -0.445514         -0.179228   
4               -0.746661            -0.164944         -0.179228   

   BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  BASEL_temp_max  \
0       -0.902918        -0.528623       -0.845652       -0.478356   
1       -0.810126        -0.582946       -0.462450       -0.569988   
2       -1.065304        -0.257010       -0.186545       -0.592896   
3       -0.114186        -0.555784       -0.385810       -0.512718   
4        0.187388        -1.003946       -1.075573       -1.039603   

   BELGRADE_cloud_cover  BELGRADE_humidity  BELGRADE_pressure  \
0             -1.206433           0.905270           0.321771   
1              0.652846           1.119118           0.016838   
2              0.652846           0.620141           0.109644   
3              1.396557           1.760659           1.289601   
4              1.396557           2.188353           1.528244   

   BELGRADE_global_radiation  BELGRADE_precipitation  BELGRADE_sunshine  \
0                  -0.743149               -0.371315           0.283983   
1                  -1.418387               -0.371315          -1.333204   
2                  -0.968228               -0.371315          -0.524611   
3                  -1.418387               -0.371315          -1.333204   
4                  -1.418387               -0.253655          -1.333204   

   BELGRADE_temp_mean  BELGRADE_temp_min  BELGRADE_temp_max  \
0           -1.016876          -1.220210          -0.949203   
1           -1.107669          -0.818702          -1.298119   
2           -1.084971          -1.168403          -1.098738   
3           -1.209812          -1.362681          -1.437685   
4           -1.209812          -1.012980          -1.457623   

   BUDAPEST_cloud_cover  BUDAPEST_humidity  BUDAPEST_pressure  \
0              -0.25524          -0.010098          -0.005876   
1              -0.25524          -0.010098          -0.005876   
2              -0.25524          -0.010098          -0.005876   
3              -0.25524          -0.010098          -0.005876   
4              -0.25524          -0.010098          -0.005876   

   BUDAPEST_global_radiation  BUDAPEST_precipitation  BUDAPEST_sunshine  \
0                  -1.118537               -0.321868          -0.773378   
1                  -1.415938                0.371622          -1.344065   
2                  -1.278676               -0.344984          -1.195190   
3                  -1.404499               -0.344984          -1.344065   
4                  -1.404499               -0.344984          -1.344065   

   BUDAPEST_temp_mean  BUDAPEST_temp_min  BUDAPEST_temp_max  \
0           -1.099163          -1.119431          -1.136839   
1           -1.110927          -0.878727          -1.341204   
2           -1.063873          -0.838610          -1.116403   
3           -1.146217          -1.012452          -1.208367   
4           -1.087400          -0.918844          -1.116403   

   DEBILT_cloud_cover  DEBILT_wind_speed  DEBILT_humidity  DEBILT_pressure  \
0            0.713782           2.781340         0.357839        -1.232471   
1            1.176060           0.446070         0.866945        -0.987868   
2            0.251504          -0.202616         1.0

In [18]:
# Concatenate the DATE and MONTH columns back to the scaled dataframe

df_scaled_final = pd.concat([date_month_columns, df_scaled], axis=1)

In [19]:
df_scaled_final

DATE  MONTH  BASEL_cloud_cover  BASEL_wind_speed  BASEL_humidity  \
0      19600101      1           0.660514          -0.02793        0.826097   
1      19600102      1           0.244897          -0.02793        0.735760   
2      19600103      1           1.076130          -0.02793        1.277781   
3      19600104      1          -1.001953          -0.02793        1.458455   
4      19600105      1           0.244897          -0.02793        1.729466   
...         ...    ...                ...               ...             ...   
22945  20221027     10          -1.833187          -0.02793        0.284075   
22946  20221028     10           0.244897          -0.02793        0.103401   
22947  20221029     10          -0.586336          -0.02793        0.013064   
22948  20221030     10          -0.170720          -0.02793        0.374412   
22949  20221031     10          -0.170720          -0.02793        0.735760   

       BASEL_pressure  BASEL_global_radiation  BASEL_precipitation  \
0           -0.001949               -1.101066            -0.265148   
1           -0.001949               -1.058108             1.658760   
2           -0.001949               -1.251420             0.155707   
3           -0.001949               -0.821838            -0.445514   
4           -0.001949               -0.746661            -0.164944   
...               ...                     ...                  ...   
22945        1.037367               -0.005632            -0.004619   
22946        0.976231               -0.005632            -0.004619   
22947        0.716402               -0.005632            -0.004619   
22948        0.487141               -0.005632            -0.004619   
22949        0.196744               -0.005632            -0.004619   

       BASEL_snow_depth  BASEL_sunshine  BASEL_temp_mean  BASEL_temp_min  \
0             -0.179228       -0.902918        -0.528623       -0.845652   
1             -0.179228       -0.810126        -0.582946       -0.462450   
2             -0.179228       -1.065304        -0.257010       -0.186545   
3             -0.179228       -0.114186        -0.555784       -0.385810   
4             -0.179228        0.187388        -1.003946       -1.075573   
...                 ...             ...              ...             ...   
22945         -0.179228        0.720943         0.747959        0.779123   
22946         -0.179228        0.187388         0.856604        1.223637   
22947         -0.179228        0.349774         0.856604        1.039701   
22948         -0.179228        0.280180         0.680055        0.809780   
22949         -0.179228       -0.322968         0.422023        0.549202   

       BASEL_temp_max  BELGRADE_cloud_cover  BELGRADE_humidity  \
0           -0.478356             -1.206433           0.905270   
1           -0.569988              0.652846           1.119118   
2           -0.592896              0.652846           0.620141   
3           -0.512718              1.396557           1.760659   
4           -1.039603              1.396557           2.188353   
...               ...                   ...                ...   
22945        0.724314             -0.834577          -0.021400   
22946        0.781585             -1.578288          -0.021400   
22947        0.838855             -0.834577          -0.021400   
22948        0.689952             -1.206433          -0.021400   
22949        0.472326             -0.090866          -0.021400   

       BELGRADE_pressure  BELGRADE_global_radiation  BELGRADE_precipitation  \
0               0.321771                  -0.743149               -0.371315   
1               0.016838                  -1.418387               -0.371315   
2               0.109644                  -0.968228               -0.371315   
3               1.289601                  -1.418387               -0.371315   
4               1.528244                  -1.418387               -0.253655   
...                  ...                 

## 4. Exporting Scaled Dataframe

In [20]:
df_scaled_final.to_csv(os.path.join(path, 'dataset_scaled.csv'), index=False)